presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,424 @@
+                    && ((ComparisonExpression) expression).getRight() instanceof SymbolReference;
+        }
+
+        private static JoinNode.EquiJoinClause toEquiJoinClause(ComparisonExpression equality, Set<Symbol> leftSymbols)
+        {
+            Symbol leftSymbol = Symbol.from(equality.getLeft());
+            Symbol rightSymbol = Symbol.from(equality.getRight());
+            JoinNode.EquiJoinClause equiJoinClause = new JoinNode.EquiJoinClause(leftSymbol, rightSymbol);
+            return leftSymbols.contains(leftSymbol) ? equiJoinClause : equiJoinClause.flip();
+        }
+
+        private JoinEnumerationResult setJoinNodeProperties(JoinNode joinNode)
+        {
+            // TODO avoid stat (but not cost) recalculation for all considered (distribution,flip) pairs, since resulting relation is the same in all case
+
+            List<JoinEnumerationResult> possibleJoinNodes = new ArrayList<>();
+            FeaturesConfig.JoinDistributionType joinDistributionType = getJoinDistributionType(session);
+
+            if (isAtMostScalar(joinNode.getRight(), lookup)) {
+                JoinNode node = joinNode.withDistributionType(REPLICATED);
+                return new JoinEnumerationResult(costProvider.getCumulativeCost(node), Optional.of(node));
+            }
+            if (isAtMostScalar(joinNode.getLeft(), lookup)) {
+                JoinNode node = joinNode.flipChildren().withDistributionType(REPLICATED);
+                return new JoinEnumerationResult(costProvider.getCumulativeCost(node), Optional.of(node));
+            }
+
+            if (joinDistributionType.canRepartition() && !joinNode.isCrossJoin()) {
+                JoinNode node = joinNode.withDistributionType(PARTITIONED);
+                possibleJoinNodes.add(new JoinEnumerationResult(costProvider.getCumulativeCost(node), Optional.of(node)));
+                node = node.flipChildren();
+                possibleJoinNodes.add(new JoinEnumerationResult(costProvider.getCumulativeCost(node), Optional.of(node)));
+            }
+            if (joinDistributionType.canReplicate()) {
+                JoinNode node = joinNode.withDistributionType(REPLICATED);
+                possibleJoinNodes.add(new JoinEnumerationResult(costProvider.getCumulativeCost(node), Optional.of(node)));
+                node = node.flipChildren();
+                possibleJoinNodes.add(new JoinEnumerationResult(costProvider.getCumulativeCost(node), Optional.of(node)));
+            }
+            if (possibleJoinNodes.stream().anyMatch(result -> result.cost.hasUnknownComponents())) {
+                return UNKNOWN_COST_RESULT;
+            }
+            return resultComparator.min(possibleJoinNodes);
+        }
+    }
+
+    @VisibleForTesting
+    static class JoinEnumerationResult
+    {
+        static final JoinEnumerationResult UNKNOWN_COST_RESULT = new JoinEnumerationResult(UNKNOWN_COST, Optional.empty());
```

Comment:
Out of the file encapsulation is ensured by the `JoinEnumerationResult` visibility level. Although it is not supported by the compiler, we used to set inner class fields/methods to be public, to indicate that those are used from outside of the inner class. Thus i suggest to make both constants to be `public`.

Reply:
`public` here triggers a warning in IntelliJ (`Java | Declaration redundancy | Declaration access can be weaker`)

Reply:
I usually disable this warnings. Similarly as i do for the `Optional` as a field ones.

Reply:
I don't think that check make sense. Intellij has other checks that don't play well with our coding style. For instance there is a check that complains if you declare a field as `Optional`.